src/agent/task_manager/src/task_manager.cpp

```diff
@@ -0,0 +1,91 @@
+#include <task_manager.hpp>
+
+#include <logger.hpp>
+
+#include <boost/asio/co_spawn.hpp>
+#include <boost/asio/detached.hpp>
+#include <boost/asio/post.hpp>
+
+#include <memory>
+#include <utility>
+
+TaskManager::TaskManager()
+    : m_work(m_ioContext)
+{
+}
+
+void TaskManager::Start(size_t numThreads)
+{
+    Stop();
+
+    for (size_t i = 0; i < numThreads; ++i)
+    {
+        m_threads.emplace_back([this]() { m_ioContext.run(); });
+    }
+}
+
+void TaskManager::Stop()
+{
+    m_ioContext.stop();
+
+    for (std::thread& thread : m_threads)
+    {
+        if (thread.joinable())
+        {
+            thread.join();
+        }
+    }
+    m_threads.clear();
+    m_ioContext.reset();
+}
+
+void TaskManager::EnqueueTask(std::function<void()> task, const std::string& taskID)
+{
+    if (++m_numEnqueuedThreads > m_threads.size() - 1) // -1 to account for the coroutines
+    {
+        LogError("Enqueued more threaded tasks than available threads");
+    }
+
+    auto taskWithThreadCounter = [this, taskID, task = std::move(task)]() mutable
+    {
+        try
+        {
+            task();
+        }
+        catch (const std::exception& e)
+        {
+            LogError("{} task threw an exception: {}", taskID.empty() ? "Anonymous" : taskID, e.what());
```

Comment:
```suggestion
            LogError("{} task threw an exception: {}. Exiting...", taskID.empty() ? "Anonymous" : taskID, e.what());
```
Or maybe relaunch the task. What do you think?

Reply:
I wouldn't try to relaunch the task because we don't know what the task is, besides how many retries should we attempt for a failing task? I think it raises too many questions to do it without research.